### PR TITLE
[dynamo] "TorchDynamo Cache Lookup" event: use C++ api

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108462
* __->__ #108436

**Background**: "TorchDynamo Cache Lookup" events appear in traces to indicate a dynamo cache lookup; it's useful to check when cache lookups are taking a long time. To add a profiler event, one can use the `torch.profiler.record_function` context manager, or the C++ equivalent. Previously, the python version was used; first, when the profiler was enabled, callbacks for record_function_enter and record_function_exit were registered; then those would be called before and after every cache lookup.

**This PR**: Instead of calling the python bindings for `torch.profiler.record_function`, directly call the C++ implementation. This simplifies a lot of the code for binding C/C++. It also improves performance; previously there was a lot of overhead in the "TorchDynamo Cache Lookup" event, making the event artificially take a long time. After this change the events now appear shorter, because there's less overhead in starting/stopping the event: in other words, the profiler no longer distorts the results as much.

**Performance results**:
I ran using the script below on a cpu-only 1.6GHz machine. I report the median time (from 100 measurements) of a "TorchDynamo Cache Lookup" event before and after this PR. I think it is reasonable to consider the difference to be due to a reduction in overhead.

<details>

<summary>Benchmarking script</summary>

```python
def fn(x, y):
    return (x * y).relu()

a, b = [torch.rand((4, 4), requires_grad=True) for _ in range(2)]

opt_fn = torch.compile(fn)

opt_fn(a, b)
opt_fn(a, b)

with torch.profiler.profile() as prof:
    opt_fn(a, b)
```

</details>

Median before PR: 198-228 us (median of 100, measured 5 times)
Median after PR: 27us

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### CODE REVIEW

File: torch/csrc/dynamo/cpp_shim.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct _PytorchRecordFunctionState {
+  void* guard;
+} _PytorchRecordFunctionState;
+
```
Comment (anijain2305) [inline torch/csrc/dynamo/cpp_shim.h:10]: Can you just declare the struct here and forward declare the `guard` part in cpp_shim.cpp?

```
struct _PytorchRecordFunctionState;
```

**Review by anijain2305 — APPROVED** [review]
LGTM. Just a minor comment on the shim wrapper to see if `void* guard` can be avoid in header file.

File: torch/csrc/dynamo/cpp_shim.h (hunk 1 above)
Comment (davidberard98) [reply torch/csrc/dynamo/cpp_shim.h]: @anijain2305 I had to change the return type to a `_PytorchRecordFunctionState*`. Could you do a sanity check?

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**